### DISCUSSION TITLE
modelling quadratic effects: specification and discrete pars plot

### DISCUSSION THREAD

**Discussion by ballardtj** [cdriveraus/ctsem#55] [Q&A]
Hi Charles,

Thanks again for this amazing package! It's really one-of-a-kind and has been a real game changer for me being able to easily do these more complex analyses.

I had a few questions about modelling quadratic effects. For simplicity, let's assume I have a model with two latent variables--var1 and var2 where:

change_in_var1 = var1 x auto1 + var2 x cross_1_on_2 + eps1

change_in_var2 = var2 x auto2 + var1 x cross_2_on_1 + eps2

and...

cross_2_on_1 = beta0 + beta1 x var1

The model above assumes that the effect of var1 on the change_in_var2 depends itself on var1

Question 1) I want to allow beta0 and beta1 to vary across individuals. I understand how to make this happen. But do I also need to make cross_2_on_1 itself vary across individuals? I would have thought variability in cross_2_on_1 would be implied by variability in beta0 and/or beta1, but I get different results depending on whether I just let the betas vary across individuals versus when I additionally allow cross_2_on_1 to vary.

Question 2) When plotting the results of this model with the ctDiscretePars function, does the cross_2_on_1  effect shown take into account the fact that the effect also changes as a function of var1? Is it marginalised across var1 somehow? Or can this plot not really be interpreted when the cross effects themselves change dynamically?

Question 3) Related to the above -- the plot generated by the ctDiscretePars plot shows two cross effects: var1.var2 and var2.var1. Which one is which? Is var1.var2 the effect of var1 on var2 or is it var1 regressed on var2?

Thanks!

Tim

**Comment by cdriveraus** [comment]
Thanks for the nice words Tim :)

1) can you post the code? From what you describe there is no specific parameter for cross2on1 so no way to make it vary except by component parameters, so something is wrong or confused somewhere. 

2) Ok i really need to update such functions to at least point out that they only describe the linear effects, or hopefully find time to somehow represent the nonlinearity. 

3) the column affects the row, and standard marrix indexing of row, column is used. So 2nd listed variable is the cause.

**Reply by ballardtj** [reply]
Hi Charles,

Thanks for the quick reply (and apologies for my slow one). Here's my code:

`ctmodel = ctModel(`
  `type='stanct',`
  `Tpoints = 20,`
  `LAMBDA=diag(2),`
  `n.latent = 2,`
  `n.manifest = 2,`
  `manifestNames = c("work_hours_scaled","mh"),`
  `latentNames = c("HPW","MH"),`
  `id = "subject",`
  `time = "time",`
  `DIFFUSION = matrix(c("sigma_change_hpw",0,0,"sigma_change_mh"),nrow=2),`
  `DRIFT = matrix(c("tau_hpw","mh_on_hpw + mh_on_hpw2 * HPW","hpw_on_mh","tau_mh"),nrow=2),`
  `CINT = c("mu_hpw","mu_mh"),`
  `T0MEANS = c("init_hpw_mean","init_mh_mean"),`
  `T0VAR = matrix(c("init_hpw_sd",0,0,"init_mh_sd"),nrow=2),`
  `MANIFESTVAR = matrix(c("hpw_error_sd",0,0,"mh_error_sd"),ncol=2),`
  `MANIFESTMEANS = c(0,0),`
  `TRAITVAR = NULL,`
  `T0TRAITEFFECT = NULL,`
  `PARS = c("mh_on_hpw","mh_on_hpw2")`
`)`

Here's the output from `ctmodel$pars`

|  | matrix        | row | col | param                     | value | transform                            | indvarying | sdscale |
| -- |---------------|-----|-----|---------------------------|-------|--------------------------------------|------------|---------|
| 1 | T0MEANS       | 1   | 1   | init_hpw_mean             | NA    | 10 * param                           | TRUE       | 1       |
| 2 | T0MEANS       | 2   | 1   | init_mh_mean              | NA    | 10 * param                           | TRUE       | 1       |
| 3 | LAMBDA        | 1   | 1   | <NA>                      | 1     | <NA>                                 | FALSE      | NA      |
| 4 | LAMBDA        | 1   | 2   | <NA>                      | 0     | <NA>                                 | FALSE      | NA      |
| 5 | LAMBDA        | 2   | 1   | <NA>                      | 0     | <NA>                                 | FALSE      | NA      |
| 6 | LAMBDA        | 2   | 2   | <NA>                      | 1     | <NA>                                 | FALSE      | NA      |
| 7 | DRIFT         | 1   | 1   | tau_hpw                   | NA    | -(1e-06 + 2 * log1p_exp(-(2 * param))) | FALSE      | 1       |
| 8 | DRIFT         | 1   | 2   | hpw_on_mh                 | NA    | param                                | FALSE      | 1       |
| 9 | DRIFT         | 2   | 1   | mh_on_hpw + mh_on_hpw2 * HPW | NA    | <NA>                                 | FALSE      | 1       |
| 10 | DRIFT         | 2   | 2   | tau_mh                    | NA    | -(1e-06 + 2 * log1p_exp(-(2 * param))) | FALSE      | 1       |
| 11 | DIFFUSION     | 1   | 1   | sigma_change_hpw          | NA    | 10 * log1p_exp(2 * param) + 1e-10    | FALSE      | 1       |
| 12 | DIFFUSION     | 1   | 2   | <NA>                      | 0     | <NA>                                 | FALSE      | NA      |
| 13 | DIFFUSION     | 2   | 1   | <NA>                      | 0     | <NA>                                 | FALSE      | NA      |
| 14 | DIFFUSION     | 2   | 2   | sigma_change_mh           | NA    | 10 * log1p_exp(2 * param) + 1e-10    | FALSE      | 1       |
| 15 | MANIFESTVAR   | 1   | 1   | hpw_error_sd              | NA    | 1e-10 + 5 * log1p_exp(2 * param)     | FALSE      | 1       |
| 16 | MANIFESTVAR   | 1   | 2   | <NA>                      | 0     | <NA>                                 | FALSE      | NA      |
| 17 | MANIFESTVAR   | 2   | 1   | <NA>                      | 0     | <NA>                                 | FALSE      | NA      |
| 18 | MANIFESTVAR   | 2   | 2   | mh_error_sd               | NA    | 1e-10 + 5 * log1p_exp(2 * param)     | FALSE      | 1       |
| 19 | MANIFESTMEANS | 1   | 1   | <NA>                      | 0     | <NA>                                 | FALSE      | NA      |
| 20 | MANIFESTMEANS | 2   | 1   | <NA>                      | 0     | <NA>                                 | FALSE      | NA      |
| 21 | CINT          | 1   | 1   | mu_hpw                    | NA    | 10 * param                           | TRUE       | 1       |
| 22 | CINT          | 2   | 1   | mu_mh                     | NA    | 10 * param                           | TRUE       | 1       |
| 23 | T0VAR         | 1   | 1   | init_hpw_sd               | NA    | 1e-10 + 5 * log1p_exp(2 * param)     | FALSE      | 1       |
| 24 | T0VAR         | 1   | 2   | <NA>                      | 0     | <NA>                                 | FALSE      | NA      |
| 25 | T0VAR         | 2   | 1   | <NA>                      | 0     | <NA>                                 | FALSE      | NA      |
| 26 | T0VAR         | 2   | 2   | init_mh_sd                | NA    | 1e-10 + 5 * log1p_exp(2 * param)     | FALSE      | 1       |
| 27 | PARS          | 1   | 1   | mh_on_hpw                 | NA    | param                                | FALSE      | 1       |
| 28 | PARS          | 2   | 1   | mh_on_hpw2                | NA    | param                                | FALSE      | 1       |

I wanted to set mh_on_hpw and mh_on_hpw2 (equivalent to beta1 and beta2 in my comment above) to vary across individuals. TO do this, I added:

`ctmodel$pars$indvarying[27] <- TRUE #set linear effect of hpw on mh to be random`
`ctmodel$pars$indvarying[28] <- TRUE #set quadratic effect of hpw on mh to be random`

My question is what in this context is meant by `ctmodel$pars$indvarying[9]`? The default of `FALSE` would appear to suggest that `mh_on_hpw + mh_on_hpw2 * HPW` does not vary across individuals, which it should given that I've specified `mh_on_hpw` and `mh_on_hpw2` to vary across individuals. Is `ctmodel$pars$indvarying[9]` overridden in this scenario? Or do I need to manually set `ctmodel$pars$indvarying[9] <- TRUE`?

I hope this makes sense!

And thank you for clarifying 2) and 3) in my original post.

Cheers,
Tim

**Comment by cdriveraus** [comment]
I see. Basically, row 9 of $pars is now a deterministic calculation, so behaves (in this sense at least) similar to if you'd specified a fixed value -- the irrelevant aspects (indvarying, sdscale, any covariate effects) are ignored. You can see what's happening in these regards a bit more clearly if you use the ctModelLatex function to view the model (note the subject parameter distribution now that I set the extra pars to have random effects individual variation). 
```
ctmodel = ctModel(
  type='stanct',
  Tpoints = 20,
  LAMBDA=diag(2),
  n.latent = 2,
  n.manifest = 2,
  manifestNames = c("work_hours_scaled","mh"),
  latentNames = c("HPW","MH"),
  id = "subject",
  time = "time",
  DIFFUSION = matrix(c("sigma_change_hpw",0,0,"sigma_change_mh"),nrow=2),
  DRIFT = matrix(c("tau_hpw","mh_on_hpw + mh_on_hpw2 * HPW","hpw_on_mh","tau_mh"),nrow=2),
  CINT = c("mu_hpw","mu_mh"),
  T0MEANS = c("init_hpw_mean","init_mh_mean"),
  T0VAR = matrix(c("init_hpw_sd",0,0,"init_mh_sd"),nrow=2),
  MANIFESTVAR = matrix(c("hpw_error_sd",0,0,"mh_error_sd"),ncol=2),
  MANIFESTMEANS = c(0,0),
  TRAITVAR = NULL,
  T0TRAITEFFECT = NULL,
  PARS = c("mh_on_hpw||TRUE","mh_on_hpw2||TRUE")
)
ctModelLatex(ctmodel)
```
![image](https://github.com/cdriveraus/ctsem/assets/4108549/cdb7b245-083d-455c-bee3-669c01d6eb38)

**Reply by ballardtj** [reply]
Ah ok that makes sense. So no need to change anything in Row 9.

Thank you so much!
